File: src/crawlers/youjizz.py
from src import helpers
from src.crawlermixin import CrawlerMixin


class YoujizzCrawler(CrawlerMixin):
    site_name = 'Youjizz'
    site_url = 'https://www.youjizz.com'
    site_favicon_url = site_url + '/favicon.ico'

    crawler_entry_point = '/newest-clips/300.html'
    crawler_selectors = dict(
        prev_page='(//*[@id="pagination"]/span/preceding-sibling::a/@href)[last()]',

        video=dict(
            title='//*[contains(@class, "video-title")]/a/text()',
            duration='//*[contains(@class, "time")]/text()',
            url='//*[contains(@class, "video-title")]/a/@href',
            thumbnail_url='//a[@class="frame"]/img[contains(@class, "lazy")]/@data-original',
        ),

        video_details=dict(
            tags='//*[contains(@class, "tag-links desktop-only")]/ul/li/a/text()',
        )
    )

    def crawl_convert_video_duration_to_seconds(self, duration):
        if not duration or duration == 'N/A' or duration == 'None':
            return 0

        return helpers.hms_to_s(duration)

File: src/sites/youjizz.ts
import * as sec from 'sec';
import { Site } from '../Site';
import { resolve as resolveUrl } from 'url';

class YouJizzSite extends Site {
  public getUrl(): string {
    return 'https://www.youjizz.com';
  }

  public getName(): string {
    return 'YouJizz';
  }

  public getFavicon(): string {
    return 'https://www.youjizz.com/favicon.ico';
  }

  public getEntryPoint(): string {
    return 'https://www.youjizz.com/newest-clips/2.html';
  }

  public getFields(): ISiteFields {
    return {
      previousPage: {
        selector:
          '(//*[contains(@class, "pagination")]/li[contains(@class, "active")]/preceding-sibling::li/a/@href)[last()]',
        normalizer: (previousPage: string) => String(resolveUrl(this.getUrl(), previousPage)),
      },
      videosUrl: {
        selector: '//*[contains(@class, "video-title")]/a/@href',
        normalizer: (urls: string[]) => urls.map(url => String(resolveUrl(this.getUrl(), url))),
      },
      videosThumbnailUrl: {
        selector: '//*[contains(@class, "video-item")]/div/a/img[contains(@class, "img-responsive")]/@data-original',
        normalizer: (thumbnailsUrl: string[]) => thumbnailsUrl.map(url => String(resolveUrl(this.getUrl(), url))),
      },
      video: {
        title: {
          selector: '//title/text()',
        },
        duration: {
          selector: '(//*[contains(@class, "video-info")]/div[3]/text())[last()]',
          normalizer: (duration: string) => sec(duration),
        },
        tags: {
          selector: '(//*[contains(@class, "tag-links")])[last()]/ul/li[contains(@class, "red-li")]/a/text()',
        },
      },
    };
  }
}

export { YouJizzSite };
